### ISSUE TITLE
allow filtering item search results by a contributor's actor id and role

### ISSUE THREAD

**Issue by stefanprobst** [SSHOC/sshoc-marketplace-backend#486]
some consumers of the sshoc marketplace api need to filter the `/api/item-search` response by a specific institutional affiliation, i.e. a specific `contributor`.

one example is the tools and services catalogue on the dariah.eu website (https://www.dariah.eu/tools-services/tools-and-services/), similar listings are in planning stage.

filtering item search results by a specific contributor's `actor.id` and `role.code` is currently rather cumbersome. how difficult would it be to add such filtering capabilities to `/api/item-search`?

---

i am aware i can get items associated with an actor via `/api/actors/:id?items=true`, however this does not provide information about the actor's role, so it would require separate requests for each search result to figure out the role and use it for filtering.

**Comment by stefanprobst** [comment]
@mkrzmr could we include this issue and #532 in the Q1 milestone? it's required for the new dariah.eu website.

**Comment by pWoz** [comment]
I suggest to do the following:

1. Add two new fields to the Solr schema. First one for `contributor actor id` and second for `contributor role`.
2. Modify the item-search endpoint so it will search on these new fields too. Currently search is done on the following fields:
`persistent_id,external_identifier,label_text_en,description_text_en,keyword_text,contributor_text,label_text,description_text`
3. To make it work we will have to re-index all the data in Solr (to add values to these two new fields);
4. Synchronization between DB and Solr will be slightly more difficult in this case.  Every time label for actor role will change we will have to re-index all documents containing this role.
5. It means that there will be fields in the Solr documents like that:

```
"contributor":["CESSDA","secord"],
"contributor_actor_id":["3","4"],
"contributor_actor_role_label":["Provider", "Funder"],
```

6. It also means that some direct Solr queries may give false positive results (for example: `contributor:"CESSDA" AND contributor_actor_id:4`). But in our case it doesn't seem to be an issue.


What do you think @mkrzmr , @KlausIllmayer ?

**Comment by stefanprobst** [comment]
@mkrzmr @KlausIllmayer can you comment on this?

**Comment by mkrzmr** [comment]
Last week we agreed to implement @pWoz suggestion and deploy to prod asap

**Comment by pWoz** [comment]
Is `contributor's actor id` what we want to have in the index @stefanprobst ? 
It is just integer internally generated by MP. It doesn't make sense to make it searchable in my opinion.

https://sshoc-marketplace-api.acdh-dev.oeaw.ac.at/swagger-ui/index.html#/actor-controller/getActors

**Comment by stefanprobst** [comment]
the main usecase is to be able to search items filtered by a contributor's actor id and role.

the DARIAH-knowledge-base stores the sshoc-marketplace actor id for DARIAH national consortia and working groups in its own database, and needs to be able to list any items stored in sshoc marketplace for that actor id

**Comment by stefanprobst** [comment]
i guess an alternative could be to extend the `/actors/:id?items=true` endpoint with info about contributors and their role.

**Comment by pWoz** [comment]
It makes more sense if actor ids are stored somewhere outside the MP. 

I added these two fields to Solr index and slightly modified search mechanism. Everything is in the PR.

Docker image with Solr have to be rebuild while deploying this version of MP.

**Comment by stefanprobst** [comment]
@pWoz i may have misunderstood your proposal. am i reading [this](https://github.com/SSHOC/sshoc-marketplace-backend/blob/28a9d7c4d456765aaffeec35b7471f95e67a8f63/src/test/java/eu/sshopencloud/marketplace/controllers/search/SearchControllerITCase.java#L846) right that actor/contributor search is now exposed via `q=`?

my suggestion was to add dedicated filter query params, similar to the ones we already have, like `d.status` or `f.keyword`

(we need to match actors exactly)

**Comment by KlausIllmayer** [comment]
Indeed, I guess we missed the point here - sorry. So it is crucial to have a way to search explicit for an actor id and actor role at the `/api/item-search`-endpoint. As an example, you like to have something like this: Give me all items where the actor is DARIAH-EU - having the id '3403' - in the role as curator - having the code 'curator'. The query would then look like `/api/item-search?c.actor-id=3403&c.actor-role=curator` (just as an example) with the additional possibility to search with `q=` in the result set.
@pWoz What do you think about this? Can this be implemented? And would it be something to use with help of an c-argument (= contributor) similar to the current f-argument (= facet) or the d-argument?

**Comment by pWoz** [comment]
Actually it is possible to be done using implementation that I did. For example sending the following query:

`http://localhost:8080/api/item-search?q=provider&d.contributor_actor_id=1`

will result in adding `fq=contributor_actor_id:1` into the Solr query. So result will contain only items with contributor id equals to 1 (the same as it is done for other parameters with `d` prefix). 

But there are two issues in here:
- it seems that `d` prefix was used intentionally to indicate that it is related with dynamic item properties. Am I correct (source code is not very talkative about intentions/vision of the implementator)?
- name of the parameters (in this case `contributor_actor_id` and `contributor_actor_role_code_text`) have to be the same as fields in Solr, they are not beautified as for properties. 

So, unless you decide otherwise, I will implement new prefix `c` and two new, simplified, keys (mentioned by you: `c.actor-role` and `c.actor-id`) that will be translated to two `fq` parameters on the Solr side.

**Comment by stefanprobst** [comment]
yes afaik `d` was meant for "dynamic"

**Comment by pWoz** [comment]
I modified the source code based on your suggestions. 
Now it is possible to add two new parameters to the search requests that will be used in Solr `fq` for the fields related with contributor. 

For example:
`/api/item-search?c.actor_id=5` 
`/api/item-search?c.actor_role=author`
should now return appropriate results. 

Let me know if it is now you expect.

Docker image with Solr have to be recreated because I changed name and type of one field.

**Comment by stefanprobst** [comment]
> /api/item-search?c.actor_id=5
> /api/item-search?c.actor_role=author

perfect, thanks! i guess combining them in `/api/item-search?c.actor_id=5&c.actor_role=author` also works?

**Comment by stefanprobst** [comment]
@mkrzmr where was this completed?

**Comment by KlausIllmayer** [comment]
We are still working on it, I'm reopening it.

**Comment by mkrzmr** [comment]
https://sshoc-marketplace-api-stage.acdh-dev.oeaw.ac.at/api/item-search?c.actor_id=1234&c.actor_role=author Works on stage, hence I assumed it was done. But I see Klaus opened a PR

**Comment by stefanprobst** [comment]
could somebody please clarify the status of this? thx

**Comment by KlausIllmayer** [comment]
should be now ready to test on stage, sorry for the hassle.

**Comment by stefanprobst** [comment]
cool, seems to work - what's the ETA to move this to production?

**Comment by KlausIllmayer** [comment]
we need to solve #544 then it is possible to move it to production. hopefully latest at end of may, but trying to be faster. it would be also good if you can have a look at https://github.com/SSHOC/sshoc-marketplace-backend/issues/467#issuecomment-4343485949 - if this can be solved, the logfiles are more cleaner and possible side effects of broken-pipe vanish.
